clippy_lints/src/manual_checked_div.rs

```diff
@@ -0,0 +1,182 @@
+                                    ..
+                                }),
+                                ..
+                            }
+                        )
+                    ) {
+                        format!("_{lhs_ty}")
+                    } else {
+                        String::new()
+                    };
+                    let suggestion = format!("{lhs_sugg}{type_suffix}.checked_div({rhs_snip})");
+                    divisions.push((e.span, suggestion));
+
+                    ControlFlow::<(), _>::Continue(Descend::No)
+                } else if eq.eq_expr(e, divisor) {
+                    if first_use.is_none() {
+                        first_use = Some(UseKind::Other);
+                    }
+                    ControlFlow::<(), _>::Continue(Descend::Yes)
+                } else {
+                    ControlFlow::<(), _>::Continue(Descend::Yes)
+                }
+            });
+
+            if found_early_use.is_some() || first_use != Some(UseKind::Division) || divisions.is_empty() {
+                return;
+            }
+
+            let mut spans: Vec<_> = divisions.iter().map(|(span, _)| *span).collect();
+            spans.push(cond.span);
+            span_lint_and_then(
+                cx,
+                MANUAL_CHECKED_DIV,
+                MultiSpan::from_spans(spans),
+                "manual checked division",
+                |diag| {
+                    diag.multipart_suggestion("consider using `checked_div`", divisions, applicability);
+                },
+            );
+        }
+    }
+}
+
+#[derive(Copy, Clone, Eq, PartialEq)]
+enum UseKind {
+    Division,
+    Other,
+}
+
+fn divisor_from_condition<'tcx>(cond: &'tcx Expr<'tcx>) -> Option<(&'tcx Expr<'tcx>, NonZeroBranch)> {
+    let ExprKind::Binary(binop, lhs, rhs) = cond.kind else {
+        return None;
+    };
+
+    match binop.node {
+        BinOpKind::Ne | BinOpKind::Lt if is_zero(lhs) => Some((rhs, NonZeroBranch::Then)),
+        BinOpKind::Ne | BinOpKind::Gt if is_zero(rhs) => Some((lhs, NonZeroBranch::Then)),
+        BinOpKind::Eq if is_zero(lhs) => Some((rhs, NonZeroBranch::Else)),
+        BinOpKind::Eq if is_zero(rhs) => Some((lhs, NonZeroBranch::Else)),
+        _ => None,
+    }
+}
+
+fn branch_block<'tcx>(
+    then: &'tcx Expr<'tcx>,
+    r#else: Option<&'tcx Expr<'tcx>>,
+    branch: NonZeroBranch,
+) -> Option<&'tcx Block<'tcx>> {
+    match branch {
+        NonZeroBranch::Then => match then.kind {
+            ExprKind::Block(block, _) => Some(block),
+            _ => None,
+        },
+        NonZeroBranch::Else => match r#else?.kind {
+            ExprKind::Block(block, _) => Some(block),
+            _ => None,
+        },
+    }
+}
+
+fn is_zero(expr: &Expr<'_>) -> bool {
+    is_integer_literal(expr, 0)
+}
+
+fn is_integer(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
+    matches!(
+        cx.typeck_results().expr_ty(expr).peel_refs().kind(),
+        ty::Uint(_) | ty::Int(_)
+    )
+}
```

Comment:
You can't lint signed and unsigned values with the same values since `MIN / -1` also returns `None` in `checked_div`.

Reply:
The only additional case returning `None` is `MIN / -1` so you'll have to check for that in addition to the rhs being zero. You can't change the meaning of the code for a lint like this if it's enabled by default.